src/Functions/sign.cpp

```diff
@@ -0,0 +1,98 @@
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+class FunctionSign : public IFunction
+{
+private:
+    const Context & context;
+
+public:
+    static constexpr auto name = "sign";
+
+    explicit FunctionSign(const Context & context_) : context(context_) { }
+    static FunctionPtr create(const Context & context) { return std::make_shared<FunctionSign>(context); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (!isNumber(arguments[0].type))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Argument of function {} should be a number, got {}",
+                getName(),
+                arguments[0].type->getName());
+
+        ColumnWithTypeAndName compare_elem{arguments[0].column, arguments[0].type, {}};
+
+        auto greater = FunctionFactory::instance().get("greater", context);
+        auto greater_compare = greater->build(ColumnsWithTypeAndName{compare_elem, compare_elem});
+
+        if (isUnsignedInteger(arguments[0].type.get()))
+        {
+            return greater_compare->getResultType();
+        }
+
+        auto compare_type = greater_compare->getResultType();
+
+        ColumnWithTypeAndName minus_elem = {compare_type, {}};
+
+        auto minus = FunctionFactory::instance().get("minus", context);
+        auto elem_minus = minus->build(ColumnsWithTypeAndName{minus_elem, minus_elem});
+
+        return elem_minus->getResultType();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto zero_column = arguments[0].type->createColumnConstWithDefaultValue(input_rows_count);
+
+        ColumnWithTypeAndName left{arguments[0].column, arguments[0].type, {}};
+        ColumnWithTypeAndName right{zero_column, arguments[0].type, {}};
+
+        auto greater = FunctionFactory::instance().get("greater", context);
+        auto greater_compare = greater->build(ColumnsWithTypeAndName{left, right});
```

Comment:
@ucasFL  maybe it make sense to extract that in variable without creating it for each function `execute` in next lines ?

Reply:
Done!